### PR TITLE
nixos/nextcloud: write config to additional config file

### LINKED ISSUE #49783: nextcloud module: changing config doesn't propagate to the module
Data that was once config and recorded on `${dataDir}/config/config.php` doesn't get updated when the configuration changes.

This is a violation of purity. And more than that, it's very surprising when trying to install, hesitating between pgsql and sqlite, and struggling before understanding that the configuration doesn't get used anyway :p

cc @globin @flokli @eqyiel @fpletz

### PR DESCRIPTION
**Pull request by Ma27**
###### Motivation for this change

One of the main problems of the Nextcloud module is that it's currently
not possible to alter e.g. database configuration after the initial
setup as it's written by their imperative installer to a file.

After some research[1] it turned out that it's possible to override all values
with an additional config file. The documentation has been
slightly updated to remain up-to-date, but the warnings should
remain there as the imperative configuration is still used and may cause
unwanted side-effects.

Also simplified the postgresql test which uses `ensure{Databases,Users}` to
configure the database.

Fixes #49783

[1] https://github.com/NixOS/nixpkgs/issues/49783#issuecomment-483063922

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Ma27** [comment]
I checked in a VM whether config changes actually work and confirmed several cases like invalid dbpass files and deployed this to my personal Nextcloud after that to check this with an existing environment and to ensure that this doesn't break any existing setups.

However it would be great if more people could take a look at this :)

@GrahamcOfBorg test nextcloud

File: nixos/modules/services/web-apps/nextcloud.xml [hunk 1]
```diff
@@ -63,17 +65,22 @@
    are used internally to configure an HTTP server using
    <literal><link xlink:href="https://php-fpm.org/">PHP-FPM</link></literal>
    and <literal>nginx</literal>. The <literal>config</literal> attribute set is
-   used for the <literal>config.php</literal> which is used for the
-   application's configuration. <emphasis>Beware: this isn't entirely pure
-   since the config is modified by the application's runtime!</emphasis>
+   used by the imperative installer and all values are written to an additional file
```
Comment (turion) [inline nixos/modules/services/web-apps/nextcloud.xml]: One can still install nextcloud imperatively?

File: nixos/modules/services/web-apps/nextcloud.xml [hunk 2]
```diff
@@ -87,35 +94,24 @@
...
-   <link xlink:href="https://github.com/NixOS/nixpkgs/issues/49783">#49783</link>,
-   for now it's unfortunately necessary to manually work around these issues.
-  </para>
-
-  <para>
-   Right now app installation and configuration is done imperatively in the nextcloud web ui or via the <literal>nextcloud-occ</literal> command line utility.
-   You can activate auto updates for your apps via
-   <literal><link linkend="opt-services.nextcloud.autoUpdateApps.enable">services.nextcloud.autoUpdateApps</link></literal>.
-  </para>
+   and can cause unwanted side-effects!</para>
```
Comment (turion) [inline nixos/modules/services/web-apps/nextcloud.xml]: You could link to #49783 as a place for discussion in case anyone has questions.

**Comment by turion** [comment]
One thing I don't understand is whether existing nextcloud installations will migrate smoothly to your system, or whether there is manual work to be done. In case that is so, it would be good to document that somewhere. Maybe in a wiki article or in the issue #49783.

File: nixos/modules/services/web-apps/nextcloud.xml (hunk 1 above)
Comment (Ma27) [reply nixos/modules/services/web-apps/nextcloud.xml]: When deploying Nextcloud the first time, the config will be generated using their installer tool (which is an imperative tool). After that it's currently not possible to change e.g. the database configuration.

**Comment by Ma27** [comment]
> One thing I don't understand is whether existing nextcloud installations will migrate smoothly to your system, or whether there is manual work to be done.

I already deployed the change onto my system and I didn't encounter any issues. Actually this simply persists stuff like database configuration in an additional configuration file to ensure that it's actually possible to change this after the first install.

**Review by fpletz — APPROVED** [review]

**Comment by Ma27** [comment]
@fpletz @globin resolved the merge conflict. Let me know if there's anything else to fix :) 

@GrahamcOfBorg test nextcloud

**Pull request merged**